In [5]:
import pandas as pd
import numpy as np

file_path = "2016.xlsx"

raw_df = pd.read_excel(
    file_path,
    sheet_name="sheet1",
    header=None
)

raw_df.head(10)

,0,1,2,3,4,5,6,7,8,9,...,39,40,41,42,43,44,45,46,47,48
0,وضعیت دستور کارهاي زخمی انتقالی از سال های گذشته,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1,مدیریت برق شهرستان,کد امور,سال,ماه,تعداد دستورکارهای بازمربوط به تست 1,NaN,NaN,NaN,NaN,تعداد دستورکارهای باز مربوط به تست2,...,NaN,NaN,NaN,NaN,جمع تعداد دستورکارهای باز مربوط به امور,NaN,NaN,NaN,NaN,NaN
2,NaN,NaN,NaN,NaN,دردست اجرا,تهیه صورت وضعیت,صورت وضعیت نزد ستاد,صورت وضعیت نزد مالی,مجموع دستورکارهای باز بهره برداری,دردست اجرا,...,صورت وضعیت نزد مشاور,صورت وضعیت نزد ستاد,صورت وضعیت نزد مالی,مجموع دستورکارهای باز توسعه و احداث و اصلاح و ...,دردست اجرا,تهیه صورت وضعیت,صورت وضعیت نزد مشاور,صورت وضعیت نزد ستاد,صورت وضعیت نزد مالی,کل\nدستورکارهای باز
3,شهر 1,6010,1401,7,0,0,0,0,0,0,...,0,0,0,5,9,0,0,0,0,9
4,شهر 2,6001,1401,7,10,0,0,0,10,0,...,0,0,0,4,44,0,0,0,0,44
5,شهر 3,5001,1401,7,6,0,1,0,7,0,...,0,0,0,13,76,0,0,1,0,77
6,شهر 4,9001,1401,7,15,0,5,0,20,0,...,1,0,0,50,124,0,8,6,0,138
7,شهر 5,7001,1401,7,34,0,1,1,36,0,...,0,0,0,52,200,0,0,2,2,204
8,شهر 6,6005,1401,7,0,0,1,0,1,0,...,0,1,0,19,43,5,2,2,0,52
9,شهر 7,8401,1401,7,17,0,1,0,18,1,...,0,0,0,48,171,0,0,1,0,172


# Raw Data Analysis

In [10]:
raw_df.shape[0] #rows

57

In [9]:
raw_df.shape[1] #columns

49

read 3 rows of head

In [11]:
raw_df.iloc[:3, :]

,0,1,2,3,4,5,6,7,8,9,...,39,40,41,42,43,44,45,46,47,48
0,وضعیت دستور کارهاي زخمی انتقالی از سال های گذشته,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1,مدیریت برق شهرستان,کد امور,سال,ماه,تعداد دستورکارهای بازمربوط به تست 1,NaN,NaN,NaN,NaN,تعداد دستورکارهای باز مربوط به تست2,...,NaN,NaN,NaN,NaN,جمع تعداد دستورکارهای باز مربوط به امور,NaN,NaN,NaN,NaN,NaN
2,NaN,NaN,NaN,NaN,دردست اجرا,تهیه صورت وضعیت,صورت وضعیت نزد ستاد,صورت وضعیت نزد مالی,مجموع دستورکارهای باز بهره برداری,دردست اجرا,...,صورت وضعیت نزد مشاور,صورت وضعیت نزد ستاد,صورت وضعیت نزد مالی,مجموع دستورکارهای باز توسعه و احداث و اصلاح و ...,دردست اجرا,تهیه صورت وضعیت,صورت وضعیت نزد مشاور,صورت وضعیت نزد ستاد,صورت وضعیت نزد مالی,کل\nدستورکارهای باز


see rows has value 

In [12]:
for row_index in range(3):
    print(f"\nRow {row_index}:")
    for column_index, value in raw_df.iloc[row_index].items():
        if pd.notna(value):
            print(column_index, ":", value)


Row 0:
0 : وضعیت دستور کارهاي زخمی انتقالی از سال های گذشته

Row 1:
0 : مدیریت برق شهرستان
1 : کد امور
2 : سال
3 : ماه
4 : تعداد دستورکارهای بازمربوط به تست 1
9 : تعداد دستورکارهای باز مربوط به تست2
14 : تعداد دستورکارهای باز مربوط به تست3
19 : تعداد دستورکارهای باز مربوط به تست4
25 : تعداد دستورکارهای باز مربوط به تست4
31 : تعداد دستورکارهای باز مربوط به تست6
37 : تعداد دستورکارهای باز مربوط به تست7
43 :  جمع تعداد دستورکارهای باز مربوط به امور

Row 2:
4 : دردست اجرا
5 : تهیه صورت وضعیت
6 : صورت وضعیت نزد ستاد
7 : صورت وضعیت نزد مالی
8 : مجموع دستورکارهای باز بهره برداری
9 : دردست اجرا
10 : تهیه صورت وضعیت
11 : صورت وضعیت نزد ستاد
12 : صورت وضعیت نزد مالی
13 : مجموع دستورکارهای باز لوازم اندازه گیری
14 : دردست اجرا
15 : تهیه صورت وضعیت
16 : صورت وضعیت نزد ستاد
17 : صورت وضعیت نزد مالی
18 : مجموع دستورکارهای باز لوازم اندازه گیری
19 : دردست اجرا
20 : تهیه صورت وضعیت
21 : صورت وضعیت نزد مشاور
22 : صورت وضعیت نزد ستاد
23 : صورت وضعیت نزد مالی
24 : مجموع دستورکارهای باز سیم به کابل
25 : 

In [13]:
raw_df.iloc[3:, :4].head(20)

,0,1,2,3
3,شهر 1,6010,1401,7
4,شهر 2,6001,1401,7
5,شهر 3,5001,1401,7
6,شهر 4,9001,1401,7
7,شهر 5,7001,1401,7
8,شهر 6,6005,1401,7
9,شهر 7,8401,1401,7
10,شهر 8,9501,1401,7
11,شهر 9,2001,1401,7
12,شهر 10,7501,1401,7


In [18]:
basic_data = raw_df.iloc[3:, :4].copy()

basic_data.columns = [
    "city_name",
    "city_code",
    "year",
    "month"
]

basic_data.head(24)

,city_name,city_code,year,month
3,شهر 1,6010,1401,7
4,شهر 2,6001,1401,7
5,شهر 3,5001,1401,7
6,شهر 4,9001,1401,7
7,شهر 5,7001,1401,7
8,شهر 6,6005,1401,7
9,شهر 7,8401,1401,7
10,شهر 8,9501,1401,7
11,شهر 9,2001,1401,7
12,شهر 10,7501,1401,7


In [19]:
basic_data["city_name"].value_counts(dropna=False)

city_name
شهر 1                   3
شهر 2                   3
جمع به تفکیک هر واحد    3
شهر 16                  3
شهر 15                  3
شهر 14                  3
شهر 13                  3
شهر 12                  3
شهر 11                  3
شهر 10                  3
شهر 9                   3
شهر 8                   3
شهر 7                   3
شهر 6                   3
شهر 5                   3
شهر 4                   3
شهر 3                   3
کل شرکت                 3
Name: count, dtype: int64

In [20]:
summary_rows = basic_data[
    basic_data["city_name"].isin([
        "جمع به تفکیک هر واحد",
        "کل شرکت"
    ])
].copy()

city_rows = basic_data[
    ~basic_data["city_name"].isin([
        "جمع به تفکیک هر واحد",
        "کل شرکت"
    ])
].copy()

In [21]:
print("City rows:", city_rows.shape[0])
print("Summary rows:", summary_rows.shape[0])

City rows: 48
Summary rows: 6


In [22]:
print("Number of cities:", city_rows["city_name"].nunique())

print(
    "Cities:",
    sorted(city_rows["city_name"].unique())
)

print(
    "Years:",
    city_rows["year"].unique()
)

print(
    "Months:",
    sorted(city_rows["month"].unique())
)

Number of cities: 16
Cities: ['شهر 1', 'شهر 10', 'شهر 11', 'شهر 12', 'شهر 13', 'شهر 14', 'شهر 15', 'شهر 16', 'شهر 2', 'شهر 3', 'شهر 4', 'شهر 5', 'شهر 6', 'شهر 7', 'شهر 8', 'شهر 9']
Years: [1401]
Months: [7, 8, 9]


In [23]:
city_code_check = (
    city_rows[
        ["city_name", "city_code"]
    ]
    .drop_duplicates()
    .sort_values("city_code")
)

city_code_check

,city_name,city_code
11,شهر 9,2001
5,شهر 3,5001
4,شهر 2,6001
8,شهر 6,6005
3,شهر 1,6010
7,شهر 5,7001
12,شهر 10,7501
16,شهر 14,8001
13,شهر 11,8015
15,شهر 13,8101


In [24]:
city_code_validation = (
    city_rows
    .groupby("city_name")["city_code"]
    .nunique()
)

city_code_validation

city_name
شهر 1     1
شهر 10    1
شهر 11    1
شهر 12    1
شهر 13    1
شهر 14    1
شهر 15    1
شهر 16    1
شهر 2     1
شهر 3     1
شهر 4     1
شهر 5     1
شهر 6     1
شهر 7     1
شهر 8     1
شهر 9     1
Name: city_code, dtype: int64

In [25]:
if (city_code_validation == 1).all():
    print("All cities have a consistent city code.")
else:
    print("Some cities have more than one city code.")

All cities have a consistent city code.


In [26]:
data_rows = raw_df.iloc[3:, :].copy()

data_rows = data_rows[
    ~data_rows[0].isin([
        "جمع به تفکیک هر واحد",
        "کل شرکت"
    ])
].copy()

data_rows.isna().sum()

0     0
1     0
2     0
3     0
4     0
5     0
6     0
7     0
8     0
9     0
10    0
11    0
12    0
13    0
14    0
15    0
16    0
17    0
18    0
19    0
20    0
21    0
22    0
23    0
24    0
25    0
26    0
27    0
28    0
29    0
30    0
31    0
32    0
33    0
34    0
35    0
36    0
37    0
38    0
39    0
40    0
41    0
42    0
43    0
44    0
45    0
46    0
47    0
48    0
dtype: int64

In [27]:
missing_values = data_rows.isna().sum()

missing_values[
    missing_values > 0
]

Series([], dtype: int64)

In [28]:
data_rows.dtypes

0     object
1     object
2     object
3     object
4     object
5     object
6     object
7     object
8     object
9     object
10    object
11    object
12    object
13    object
14    object
15    object
16    object
17    object
18    object
19    object
20    object
21    object
22    object
23    object
24    object
25    object
26    object
27    object
28    object
29    object
30    object
31    object
32    object
33    object
34    object
35    object
36    object
37    object
38    object
39    object
40    object
41    object
42    object
43    object
44    object
45    object
46    object
47    object
48    object
dtype: object

In [29]:
for column in range(1, 49):
    data_rows[column] = pd.to_numeric(
        data_rows[column],
        errors="coerce"
    )

In [30]:
data_rows.dtypes

0     object
1      int64
2      int64
3      int64
4      int64
5      int64
6      int64
7      int64
8      int64
9      int64
10     int64
11     int64
12     int64
13     int64
14     int64
15     int64
16     int64
17     int64
18     int64
19     int64
20     int64
21     int64
22     int64
23     int64
24     int64
25     int64
26     int64
27     int64
28     int64
29     int64
30     int64
31     int64
32     int64
33     int64
34     int64
35     int64
36     int64
37     int64
38     int64
39     int64
40     int64
41     int64
42     int64
43     int64
44     int64
45     int64
46     int64
47     int64
48     int64
dtype: object

In [31]:
duplicates = data_rows.duplicated(
    subset=[1, 2, 3],
    keep=False
)

data_rows[duplicates]

,0,1,2,3,4,5,6,7,8,9,...,39,40,41,42,43,44,45,46,47,48


In [32]:
print(
    "Number of duplicate city-period rows:",
    duplicates.sum()
)

Number of duplicate city-period rows: 0


In [33]:
numeric_values = data_rows.iloc[:, 4:49]

negative_count = (numeric_values < 0).sum().sum()

print("Number of negative values:", negative_count)

Number of negative values: 0


In [34]:
monthly_totals = (
    data_rows
    .groupby(3)[48]
    .sum()
    .reset_index()
)

monthly_totals.columns = [
    "month",
    "total_open_work_orders"
]

monthly_totals

,month,total_open_work_orders
0,7,1559
1,8,1210
2,9,1106


In [35]:
monthly_totals["change"] = (
    monthly_totals["total_open_work_orders"].diff()
)

monthly_totals["change_percent"] = (
    monthly_totals["total_open_work_orders"]
    .pct_change()
    .mul(100)
    .round(2)
)

monthly_totals

,month,total_open_work_orders,change,change_percent
0,7,1559,NaN,NaN
1,8,1210,-349.0,-22.39
2,9,1106,-104.0,-8.60


جمع‌بندی اولیه تحلیل داده خام

تا اینجا به این نتایج رسیدیم:

فایل دارای 57 ردیف و 49 ستون است.
سه ردیف اول مربوط به عنوان‌ها هستند.
48 ردیف واقعی داریم.
داده مربوط به 16 شهر است.
داده‌ها مربوط به سال 1401 و ماه‌های 7، 8 و 9 هستند.
برای هر شهر و ماه فقط یک ردیف وجود دارد.
شش ردیف مربوط به جمع‌ها هستند و باید از داده‌های اصلی جدا شوند.
Dataset به‌شکل Wide است و در مرحله بعد باید به Long تبدیل شود.
مجموع دستورکارهای باز از 1559 در ماه ۷ به 1106 در ماه ۹ کاهش یافته است.
عنوان برخی دسته‌ها مثل تست 1 تا تست 7 نیاز به تعیین نام واقعی دارد.


# DATA Cleaning

In [41]:
data = raw_df.iloc[3:].copy()

data = data[
    ~data[0].isin([
        "جمع به تفکیک هر واحد",
        "کل شرکت"
    ])
].copy()

data.reset_index(drop=True, inplace=True)

print(data.shape)
data.head()

(48, 49)


,0,1,2,3,4,5,6,7,8,9,...,39,40,41,42,43,44,45,46,47,48
0,شهر 1,6010,1401,7,0,0,0,0,0,0,...,0,0,0,5,9,0,0,0,0,9
1,شهر 2,6001,1401,7,10,0,0,0,10,0,...,0,0,0,4,44,0,0,0,0,44
2,شهر 3,5001,1401,7,6,0,1,0,7,0,...,0,0,0,13,76,0,0,1,0,77
3,شهر 4,9001,1401,7,15,0,5,0,20,0,...,1,0,0,50,124,0,8,6,0,138
4,شهر 5,7001,1401,7,34,0,1,1,36,0,...,0,0,0,52,200,0,0,2,2,204


In [42]:
data.shape

(48, 49)

In [43]:
data = data.rename(columns={
    0: "city_name",
    1: "city_code",
    2: "year",
    3: "month"
})

In [45]:
data["city_name"] = data["city_name"].astype(str).str.strip()

data["city_code"] = pd.to_numeric(
    data["city_code"],
    errors="raise"
).astype(int)

data["year"] = pd.to_numeric(
    data["year"],
    errors="raise"
).astype(int)

data["month"] = pd.to_numeric(
    data["month"],
    errors="raise"
).astype(int)

In [46]:
data[
    ["city_name", "city_code", "year", "month"]
].head()

,city_name,city_code,year,month
0,شهر 1,6010,1401,7
1,شهر 2,6001,1401,7
2,شهر 3,5001,1401,7
3,شهر 4,9001,1401,7
4,شهر 5,7001,1401,7


In [47]:
activity_groups = [
    {
        "activity": "test_1",
        "columns": {
            4: "in_progress",
            5: "statement_preparation",
            6: "at_headquarters",
            7: "at_finance"
        },
        "total_column": 8
    },
    {
        "activity": "test_2",
        "columns": {
            9: "in_progress",
            10: "statement_preparation",
            11: "at_headquarters",
            12: "at_finance"
        },
        "total_column": 13
    },
    {
        "activity": "test_3",
        "columns": {
            14: "in_progress",
            15: "statement_preparation",
            16: "at_headquarters",
            17: "at_finance"
        },
        "total_column": 18
    },
    {
        "activity": "test_4_a",
        "columns": {
            19: "in_progress",
            20: "statement_preparation",
            21: "at_consultant",
            22: "at_headquarters",
            23: "at_finance"
        },
        "total_column": 24
    },
    {
        "activity": "test_4_b",
        "columns": {
            25: "in_progress",
            26: "statement_preparation",
            27: "at_consultant",
            28: "at_headquarters",
            29: "at_finance"
        },
        "total_column": 30
    },
    {
        "activity": "test_6",
        "columns": {
            31: "in_progress",
            32: "statement_preparation",
            33: "at_consultant",
            34: "at_headquarters",
            35: "at_finance"
        },
        "total_column": 36
    },
    {
        "activity": "test_7",
        "columns": {
            37: "in_progress",
            38: "statement_preparation",
            39: "at_consultant",
            40: "at_headquarters",
            41: "at_finance"
        },
        "total_column": 42
    }
]

In [49]:
measurement_columns = list(range(4, 49))

for column in measurement_columns:
    data[column] = pd.to_numeric(
        data[column],
        errors="coerce"
    )

In [50]:
print(
    "Missing measurement values:",
    data[measurement_columns].isna().sum().sum()
)

Missing measurement values: 0


In [51]:
validation_results = []

for group in activity_groups:
    status_columns = list(group["columns"].keys())
    total_column = group["total_column"]

    calculated_total = data[status_columns].sum(axis=1)
    reported_total = data[total_column]

    invalid_rows = calculated_total != reported_total

    validation_results.append({
        "activity": group["activity"],
        "invalid_rows": int(invalid_rows.sum())
    })

validation_df = pd.DataFrame(validation_results)

validation_df

,activity,invalid_rows
0,test_1,0
1,test_2,0
2,test_3,0
3,test_4_a,0
4,test_4_b,0
5,test_6,0
6,test_7,0


In [54]:
for group in activity_groups:
    status_columns = list(group["columns"].keys())
    total_column = group["total_column"]

    calculated_total = data[status_columns].sum(axis=1)
    invalid_rows = calculated_total != data[total_column]

    if invalid_rows.any():
        print("\nActivity:", group["activity"])

        display(
            data.loc[
                invalid_rows,
                ["city_name", "year", "month", total_column]
            ]
        )

In [55]:
long_records = []

for _, row in data.iterrows():

    for group in activity_groups:

        activity_name = group["activity"]

        for column_number, status_name in group["columns"].items():

            long_records.append({
                "city_name": row["city_name"],
                "city_code": row["city_code"],
                "year": row["year"],
                "month": row["month"],
                "activity": activity_name,
                "status": status_name,
                "work_order_count": row[column_number]
            })

clean_df = pd.DataFrame(long_records)

In [63]:
clean_df.head(30)

,city_name,city_code,year,month,activity,status,work_order_count
0,شهر 1,6010,1401,7,test_1,in_progress,0
1,شهر 1,6010,1401,7,test_1,statement_preparation,0
2,شهر 1,6010,1401,7,test_1,at_headquarters,0
3,شهر 1,6010,1401,7,test_1,at_finance,0
4,شهر 1,6010,1401,7,test_2,in_progress,0
5,شهر 1,6010,1401,7,test_2,statement_preparation,0
6,شهر 1,6010,1401,7,test_2,at_headquarters,0
7,شهر 1,6010,1401,7,test_2,at_finance,0
8,شهر 1,6010,1401,7,test_3,in_progress,0
9,شهر 1,6010,1401,7,test_3,statement_preparation,0


In [60]:
clean_df = clean_df.astype({
    "city_name": "string",
    "city_code": "int64",
    "year": "int64",
    "month": "int64",
    "activity": "string",
    "status": "string",
    "work_order_count": "int64"
})

In [61]:
clean_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1536 entries, 0 to 1535
Data columns (total 7 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   city_name         1536 non-null   string
 1   city_code         1536 non-null   int64 
 2   year              1536 non-null   int64 
 3   month             1536 non-null   int64 
 4   activity          1536 non-null   string
 5   status            1536 non-null   string
 6   work_order_count  1536 non-null   int64 
dtypes: int64(4), string(3)
memory usage: 84.1 KB


In [62]:
print("Clean data shape:", clean_df.shape)

Clean data shape: (1536, 7)


In [64]:
duplicate_mask = clean_df.duplicated(
    subset=[
        "city_code",
        "year",
        "month",
        "activity",
        "status"
    ],
    keep=False
)

print("Duplicate rows:", duplicate_mask.sum())

Duplicate rows: 0


In [65]:
print(
    "Missing values:",
    clean_df.isna().sum().sum()
)

print(
    "Negative counts:",
    (clean_df["work_order_count"] < 0).sum()
)

print(
    "Invalid months:",
    (~clean_df["month"].between(1, 12)).sum()
)

Missing values: 0
Negative counts: 0
Invalid months: 0


In [67]:
calculated_totals = (
    clean_df
    .groupby(
        ["city_code", "year", "month"],
        as_index=False
    )["work_order_count"]
    .sum()
    .rename(
        columns={
            "work_order_count": "calculated_total"
        }
    )
)
reported_totals = (
    data[
        ["city_code", "year", "month", 48]
    ]
    .rename(
        columns={
            48: "reported_total"
        }
    )
)

In [68]:
total_check = calculated_totals.merge(
    reported_totals,
    on=["city_code", "year", "month"],
    how="left"
)

total_check["difference"] = (
    total_check["calculated_total"]
    - total_check["reported_total"]
)

total_check.head()

,city_code,year,month,calculated_total,reported_total,difference
0,2001,1401,7,220,220,0
1,2001,1401,8,162,162,0
2,2001,1401,9,144,144,0
3,5001,1401,7,77,77,0
4,5001,1401,8,72,72,0


In [69]:
print(
    "Rows with total mismatch:",
    (total_check["difference"] != 0).sum()
)

Rows with total mismatch: 0


In [70]:
clean_df.to_csv(
    "clean_work_orders.csv",
    index=False,
    encoding="utf-8-sig"
)

In [72]:
test_df = pd.read_csv("clean_work_orders.csv")

print(test_df.shape)
test_df.head()

(1536, 7)


,city_name,city_code,year,month,activity,status,work_order_count
0,شهر 1,6010,1401,7,test_1,in_progress,0
1,شهر 1,6010,1401,7,test_1,statement_preparation,0
2,شهر 1,6010,1401,7,test_1,at_headquarters,0
3,شهر 1,6010,1401,7,test_1,at_finance,0
4,شهر 1,6010,1401,7,test_2,in_progress,0
